### PR TITLE
Improve pipeline for multilabel setup

### PR DESCRIPTION
**Pull request by kprokofi**
- [x] onecycle for SAM, improved handling of the parameters
- [x] early stopping for multilabel
- [x]  other minor fixes

### CODE REVIEW

File: scripts/default_config.py [hunk 1]
```diff
@@ -132,9 +132,10 @@ def get_default_config():
     cfg.train.gamma = 0.1  # learning rate decay multiplier
     cfg.train.first_cycle_steps = 5
     cfg.train.cycle_mult = 1.
-    cfg.train.min_lr = 1e-5
-    cfg.train.max_lr = 0.1
+    cfg.train.min_lr = [1e-5]
```
Comment (sovrasov) [inline scripts/default_config.py]: Why do you need a list here?

File: configs/multilabel_classification/mobilenetv3.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  lr_decay_factor: 1e3
+  deterministic: True
+  target_metric: test_acc
+  gamma: 0.1
+  ema:
+    enable: True
+    ema_decay: 0.9997
+
+sam:
+  enable: True
```
Comment (sovrasov) [inline configs/multilabel_classification/mobilenetv3.yml]: I think this flag introduces ambiguity, since SAM can be also specified via train.optim.
I'd propose remove this section and move rho the to train. or train.sam group.

Comment (kprokofi) [reply configs/multilabel_classification/mobilenetv3.yml]: Agree. Ok

File: scripts/default_config.py (hunk 1 above)
Comment (kprokofi) [reply scripts/default_config.py]: if we have more than one parameter group (i.e. NBD) - we may want to use different learning rates. So, the list passes to cycling lr shedulers

File: torchreid/engine/engine.py [hunk 3]
```diff
@@ -245,38 +245,21 @@ def update_lr(self, names=None, output_avg_metric=None):
...
         '''
         The function returns a pair (should_exit, is_candidate_for_best).
 
-        The function sets this checkpoint as a candidate for best if either it is the first checkpoint
-        for this LR or this checkpoint is better then the previous best.
-
-        The function sets should_exit = True if the LR is the minimal allowed
-        LR (i.e. self.lb_lr) and the best checkpoint is not changed for self.train_patience
-        epochs.
+        For raw implementation the method returns False for should_exit.
```
Comment (sovrasov) [inline torchreid/engine/engine.py]: Default implementation of the method returns ...

File: torchreid/engine/image/am_softmax.py [hunk 4]
```diff
@@ -22,6 +22,7 @@
 import copy
 
 import numpy as np
+from numpy.lib.arraysetops import isin
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py]: Seems like this import is not used

File: torchreid/engine/image/am_softmax.py [hunk 5]
```diff
@@ -74,7 +75,9 @@ def __init__(self, datamanager, models, optimizers, reg_cfg, metric_cfg, schedul
         self.mix_weight = mix_weight
         self.clip_grad = clip_grad
         self.enable_rsc = enable_rsc
-        self.enable_sam = enable_sam
+        self.enable_sam = isinstance(self.optims[self.main_model_name], SAM)
+        for model_name in self.get_model_names():
+            assert isinstance(self.optims[model_name], SAM) == self.enable_sam, "SAM must be for all models or none of them"
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py]: must be enabled

File: torchreid/optim/lr_scheduler.py [hunk 6]
```diff
@@ -16,23 +16,23 @@ def build_lr_scheduler(optimizer, lr_scheduler, base_scheduler, **kwargs):
...
-                max_epoch=1,
-                warmup=10,
-                multiplier = 10,
-                first_cycle_steps=10,
-                cycle_mult = 1.,
-                min_lr = 1e-4,
-                max_lr = 0.1,
-                patience = 5,
-                lr_decay_factor= 100):
+                    num_iter,
```
Comment (sovrasov) [inline torchreid/optim/lr_scheduler.py]: Please add one extra space to indent

File: torchreid/optim/lr_scheduler.py [hunk 7]
```diff
@@ -16,23 +16,23 @@ def build_lr_scheduler(optimizer, lr_scheduler, base_scheduler, **kwargs):
...
+                    lr_scheduler='single_step',
+                    base_scheduler=None,
+                    stepsize=1,
+                    gamma=0.1,
+                    max_epoch=1,
+                    warmup=10,
+                    multiplier=10,
+                    first_cycle_steps=10,
+                    cycle_mult=1.,
+                    min_lr=[1e-4],
```
Comment (sovrasov) [inline torchreid/optim/lr_scheduler.py]: These two params are used in `CosineAnnealingCycleRestart` only and there is no way to deduce the actual number of parameter groups from the config, so I suggest keeping them as float numbers.

File: scripts/default_config.py (hunk 1 above)
Comment (sovrasov) [reply scripts/default_config.py]: These params are actually passed to CosineAnnealingCycleRestart only. Default setting would cause a runtime error if NBD is enabled with the mentioned scheduler. For other schedulers you generate min/max lrs right inplace taking into account the real number of parameter groups.
